### PR TITLE
Feat 3.x

### PR DESCRIPTION
**Pull request by JaydenLiang**

### CODE REVIEW

File: templates/autoscale-tgw-new.vpc.template [hunk 1]
```diff
@@ -0,0 +1,962 @@
...
+            "AllowedValues": [
+                "t2.small",
+                "c5.large",
+                "c5.xlarge",
+                "c5.2xlarge",
+                "c5.4xlarge",
+                "c5.9xlarge"
+            ],
+            "ConstraintDescription": "must be a valid EC2 instance type.",
+            "Description": "Instance type for the FortiGates in the Auto Scaling group. There are t2.small and compute-optimized instances such as c4 and c5 available with different vCPU sizes and bandwidths. For more information about instance types, see https://aws.amazon.com/ec2/instance-types/."
```
[the remark this answers is not in the document]
Comment (klc178) [reply templates/autoscale-tgw-new.vpc.template]:
```suggestion
            "Description": "Instance type for the FortiGates in the Auto Scaling group. Instance types are available with different vCPU sizes and bandwidths. For more information, see https://aws.amazon.com/ec2/instance-types/."
```

File: templates/autoscale-main.template [hunk 2]
```diff
@@ -0,0 +1,1927 @@
...
+            "Description": "The ID of the private subnet 2 located in Availability Zone 2 of the selected existing VPC. This subnet will be protected by the FortiGates in the public subnet of the same AZ."
+        },
+        "PrivateSubnetRouteTable": {
+            "Type": "String",
+            "Default": "",
+            "Description": "Route table ID associated with the two private subnets."
+        },
+        "FortiGateInstanceType": {
+            "Type": "String",
+            "Description": "Instance type for the FortiGates in the Auto Scaling group. There are t2.small and compute-optimized instances such as c4 and c5 available with different vCPU sizes and bandwidths. For more information about instance types, see https://aws.amazon.com/ec2/instance-types/."
```
Comment (klc178) [inline templates/autoscale-main.template:74]:
```suggestion
            "Description": "Instance type for the FortiGates in the Auto Scaling group. Instance types are available with different vCPU sizes and bandwidths. For more information, see https://aws.amazon.com/ec2/instance-types/."
```

File: templates/autoscale-main.template [hunk 3]
```diff
@@ -0,0 +1,1805 @@
...
+        },
+        "FortiOSVersion": {
+            "Type": "String",
+            "Description": "FortiOS version."
```
[the remark this answers is not in the document]
Comment (klc178) [reply templates/autoscale-main.template:78]: Does not appear to have been applied.

File: templates/autoscale-main.template [hunk 4]
```diff
@@ -0,0 +1,1805 @@
...
+            "Default": 2,
+            "MinValue": 0,
+            "Description": "The maximum amount of time (in seconds) allowed for network latency of the FortiGate heartbeat arriving at the Autoscale handler. Minimum is 0."
+        },
+        "LoadBalancingTrafficProtocol": {
+            "Type": "String",
+            "Default": "HTTPS",
+            "AllowedValues": ["HTTP", "HTTPS", "TCP"],
+            "ConstraintDescription": "value must be chosen from the provided options.",
+            "Description": "Use this protocol to load balance traffic."
```
[the remark this answers is not in the document]
Comment (klc178) [reply templates/autoscale-main.template:221]: Does not appear to have been applied.

File: templates/autoscale-main.template [hunk 5]
```diff
@@ -0,0 +1,1805 @@
...
+        "InternalLoadBalancingOptions": {
+            "Type": "String",
+            "Default": "do not need one",
+            "AllowedValues": [
+                "add a new internal load balancer",
+                "use a load balancer specified below",
+                "do not need one"
+            ],
+            "ConstraintDescription": "must choose from the provided options.",
+            "Description": "Add an optional pre-defined load balancer to route traffic to web service in the private subnets. You can optionally use your own one or decide to not need one."
```
[the remark this answers is not in the document]
Comment (klc178) [reply templates/autoscale-main.template:247]: applied?

File: templates/autoscale-main.template [hunk 6]
```diff
@@ -0,0 +1,1805 @@
...
+        },
+        "InternalTargetGroupHealthCheckPath": {
+            "Type": "String",
+            "Default": "/",
+            "AllowedPattern": "^/[0-9a-zA-Z-/]*$",
+            "MaxLength": 1024,
+            "ConstraintDescription": "This path must begin with a '/' character, and can be at most 1024 characters in length.",
+            "Description": "Optional. The destination path for health checks. This path must begin with a '/' character, and can be at most 1024 characters in length."
```
[the remark this answers is not in the document]
Comment (klc178) [reply templates/autoscale-main.template:255]: applied?

File: templates/autoscale-main.template [hunk 7]
```diff
@@ -0,0 +1,1805 @@
...
+        },
+        "InternalLoadBalancerDnsName": {
+            "Type": "String",
+            "Default": "",
+            "Description": "Optional. Specify the DNS Name of an existing internal load balancer used to route traffic from a FortiGate to targets in a specified target group. Leave it blank if you don't use an existing load balancer."
```
[the remark this answers is not in the document]
Comment (klc178) [reply templates/autoscale-main.template:260]: applied?

File: templates/autoscale-new-vpc.template [hunk 8]
```diff
@@ -0,0 +1,1020 @@
...
+            "Description": "Instance type for the FortiGates in the Auto Scaling group. There are t2.small and compute-optimized instances such as c4 and c5 available with different vCPU sizes and bandwidths. For more information about instance types, see https://aws.amazon.com/ec2/instance-types/."
+        },
+        "FortiOSVersion": {
+            "Type": "String",
+            "Default": "6.2.3",
+            "AllowedValues": [
+                "6.2.3"
+            ],
+            "ConstraintDescription": "must be a valid FortiOS version from the selection.",
+            "Description": "FortiOS version."
```
[the remark this answers is not in the document]
Comment (klc178) [reply templates/autoscale-new-vpc.template:90]: applied?

File: templates/autoscale-new-vpc.template [hunk 9]
```diff
@@ -0,0 +1,1020 @@
...
+        "LoadBalancingTrafficProtocol": {
+            "Type": "String",
+            "Default": "HTTPS",
+            "AllowedValues": [
+                "HTTP",
+                "HTTPS",
+                "TCP"
+            ],
+            "ConstraintDescription": "value must be chosen from the provided options.",
+            "Description": "Use this protocol to load balance traffic."
```
[the remark this answers is not in the document]
Comment (klc178) [reply templates/autoscale-new-vpc.template:240]: applied?

File: templates/autoscale-new-vpc.template [hunk 10]
```diff
@@ -0,0 +1,1020 @@
...
+        "InternalLoadBalancingOptions": {
+            "Type": "String",
+            "Default": "add a new internal load balancer",
+            "AllowedValues": [
+                "add a new internal load balancer",
+                "use a load balancer specified below",
+                "do not need one"
+            ],
+            "ConstraintDescription": "must choose from the provided options.",
+            "Description": "Add an optional pre-defined load balancer to route traffic to web service in the private subnets. You can optionally use your own one or decide to not need one."
```
[the remark this answers is not in the document]
Comment (klc178) [reply templates/autoscale-new-vpc.template:266]: applied?

File: templates/autoscale-new-vpc.template [hunk 11]
```diff
@@ -0,0 +1,1020 @@
...
+        },
+        "InternalTargetGroupHealthCheckPath": {
+            "Type": "String",
+            "Default": "/",
+            "AllowedPattern": "^/[0-9a-zA-Z-/]*$",
+            "MaxLength": 1024,
+            "ConstraintDescription": "This path must begin with a '/' character, and can be at most 1024 characters in length.",
+            "Description": "Optional. The destination path for health checks. This path must begin with a '/' character, and can be at most 1024 characters in length."
```
[the remark this answers is not in the document]
Comment (klc178) [reply templates/autoscale-new-vpc.template:274]: applied?

File: templates/autoscale-new-vpc.template [hunk 12]
```diff
@@ -0,0 +1,1020 @@
...
+        },
+        "InternalLoadBalancerDnsName": {
+            "Type": "String",
+            "Default": "",
+            "Description": "Optional. Specify the DNS Name of an existing internal load balancer used to route traffic from a FortiGate to targets in a specified target group. Leave it blank if you don't use an existing load balancer."
```
[the remark this answers is not in the document]
Comment (klc178) [reply templates/autoscale-new-vpc.template:279]: applied?

File: templates/autoscale-tgw-new-vpc.template [hunk 13]
```diff
@@ -0,0 +1,963 @@
...
+            "AllowedValues": [
+                "t2.small",
+                "c5.large",
+                "c5.xlarge",
+                "c5.2xlarge",
+                "c5.4xlarge",
+                "c5.9xlarge"
+            ],
+            "ConstraintDescription": "must be a valid EC2 instance type.",
+            "Description": "Instance type for the FortiGates in the Auto Scaling group. There are t2.small and compute-optimized instances such as c4 and c5 available with different vCPU sizes and bandwidths. For more information about instance types, see https://aws.amazon.com/ec2/instance-types/."
```
Comment (klc178) [inline templates/autoscale-tgw-new-vpc.template:69]:
```suggestion
            "Description": "Instance type for the FortiGates in the Auto Scaling group. Instance types are available with different vCPU sizes and bandwidths. For more information, see https://aws.amazon.com/ec2/instance-types/."
```

File: templates/autoscale-tgw-new-vpc.template [hunk 14]
```diff
@@ -0,0 +1,963 @@
...
+        },
+        "FortiOSVersion": {
+            "Type": "String",
+            "Default": "6.2.3",
+            "AllowedValues": ["6.2.3"],
+            "ConstraintDescription": "must be a valid FortiOS version from the selection.",
+            "Description": "FortiOS version."
```
Comment (klc178) [inline templates/autoscale-tgw-new-vpc.template:76]:
```suggestion
	            "Description": "FortiOS version supported by FortiGate Autoscale for AWS."
```

File: templates/create-db-table.template [hunk 15]
```diff
@@ -0,0 +1,853 @@
...
+            }
+        },
+        "FortiAnalyzerTableArn": {
+            "Condition": "EnableFortiAnalyzer",
+            "Value": {
+                "Fn::GetAtt": ["FortiAnalyzer", "Arn"]
+            }
+        },
+        "TableArnList": {
+            "Description": "The comma-delimited list of arn of DynamoDB tables created in this stack.",
```
Comment (klc178) [inline templates/create-db-table.template:634]:
```suggestion
            "Description": "The comma-delimited list of ARNs for DynamoDB tables created in this stack.",
```

File: templates/create-hybrid-auto-scaling-group.template [hunk 16]
```diff
@@ -0,0 +1,1035 @@
...
+            "AllowedValues": [
+                "t2.small",
+                "c5.large",
+                "c5.xlarge",
+                "c5.2xlarge",
+                "c5.4xlarge",
+                "c5.9xlarge"
+            ],
+            "ConstraintDescription": "must be a valid EC2 instance type.",
+            "Description": "Instance type to launch as FortiGate On-Demand instances. There are t2.small and compute-optimized instances such as c4 and c5 available with different vCPU sizes and bandwidths. For more information about instance types, see https://aws.amazon.com/ec2/instance-types/"
```
Comment (klc178) [inline templates/create-hybrid-auto-scaling-group.template:55]:
```suggestion
            "Description": "Instance type for the FortiGates in the Auto Scaling group. Instance types are available with different vCPU sizes and bandwidths. For more information, see https://aws.amazon.com/ec2/instance-types/."
```

**Review by klc178 — CHANGES_REQUESTED** [review]

File: templates/autoscale-tgw-new-vpc.template [hunk 17]
```diff
@@ -66,14 +66,14 @@
...
-            "Description": "Instance type for the FortiGates in the Auto Scaling group. There are t2.small and compute-optimized instances such as c4 and c5 available with different vCPU sizes and bandwidths. For more information about instance types, see https://aws.amazon.com/ec2/instance-types/."
+            "Description": "Instance type for the FortiGates in the Auto Scaling group. Instance types are available with different vCPU sizes and bandwidths. For more information, see https://aws.amazon.com/ec2/instance-types/."
         },
         "FortiOSVersion": {
             "Type": "String",
             "Default": "6.2.3",
             "AllowedValues": ["6.2.3"],
             "ConstraintDescription": "must be a valid FortiOS version from the selection.",
-            "Description": "FortiOS version."
+	            "Description": "FortiOS version supported by FortiGate Autoscale for AWS."
```
Comment (JaydenLiang) [inline templates/autoscale-tgw-new-vpc.template:76]:
```suggestion
            "Description": "FortiOS version supported by FortiGate Autoscale for AWS."
```

**Review by klc178 — APPROVED** [review]
The changes in https://github.com/fortinet/fortigate-autoscale-aws/pull/1/commits/75c3d3ab774b58dd24967be72e68eba93cf72053 look good.

**Review by klc178 — APPROVED** [review]
Changes in https://github.com/fortinet/fortigate-autoscale-aws/pull/1/commits/2959085ee59cea60f31d40587c9570c4380edf72 look good.

**Review by jamie-pate — APPROVED** [review]

**Pull request merged**